src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3858,7 +3858,7 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
 
             IOperation tryDispose(IOperation value)
             {
-                Debug.Assert(value.Type == iDisposable);
+                Debug.Assert(Equals(value.Type, iDisposable));
```

Comment:
Please add a corresponding test.

Reply:
i have no idea how.  I hit this while opening roslyn in roslyn.  it's clearly busted, and this addrsses the issue (validated in roslyn and in the debugger).  If you guys can figure out a test that will make this repro, i'll add one.  But i really don't know understand what circumstances the compiler will generate a fresh instances for IDisposable vs reusing the existing one.

Reply:
Also tagging @AlekseyTs @333fred

Reply:
> But i really don't know understand what circumstances the compiler will generate a fresh instances for IDisposable vs reusing the existing one.

The only scenario I can think of is some kind of error scenario involving the IDisposable type being missing. It would be interesting to understand how could this happen in Roslyn.
Given that there is only one callsite for the ```tryDispose``` method and it is preceded with
```
            if (!resource.Type.Equals(iDisposable))
            {
                resource = ConvertToIDisposable(resource, iDisposable);
            }
```

I feel comfortable with changing the assert to use ```Equals``` method. Though I would prefer us to use the same form and the same method, i.e. ```value.Type.Equals(iDisposable)``` 
 

---
In reply to: [347547587](https://github.com/dotnet/roslyn/pull/39866#discussion_r347547587) [](ancestors = 347547587)

Reply:
> The only scenario I can think of is some kind of error scenario involving the IDisposable type being missing.

It could definitely have been that.  Whena large project loads, the references are actually being added in the background.  So it's def possible to have a period of time when some types are missing, but then become available.  

I've updated the code to match what you asked for.  Thanks!